src/helm/benchmark/scenarios/cleva_scenario.py

```diff
@@ -0,0 +1,1491 @@
+            references=references,
+            split=split,
+        )
+        return instance
+
+    def transform_code(self, data: Dict[str, Any], templates: Dict[str, Any], split: str) -> CodeInstance:
+        """
+        Similar to transform method above, transform_code converts a data point in code synthesis scenario in CLEVA
+        to a HELM CodeInstance according to a given CLEVA prompt template.
+        """
+
+        data["prompt"] = templates["input"].format(**data)
+        instance = CodeInstance(
+            input=Input(text=data["prompt"]),
+            references=[
+                CodeReference(
+                    output=Output(text=data["canonical_solution"]),
+                    test_cases=data,
+                    tags=[CORRECT_TAG],
+                )
+            ],
+            split=split,
+        )
+        return instance
+
+    def _apply_all(self, data: dict, templates: dict) -> dict:
+        """
+        This function applies the CLEVA prompt template to a data point.
+
+        Note that this is an in-place operation.
+
+        The logic is as follows:
+        1. It first maps every entry according to a set of predefined mappings in "verbalizer".
+        2. It then stringifies all fields in the given data point, including processing structured data.
+        3. It finally constructs the input string and reference strings.
+        """
+        # If we define a verbalizer, we map all fields before further processing
+        if templates.get("verbalizer", None) is not None:
+            self._mapping_all(data, templates["verbalizer"])
+
+        if "choices" in data.keys():
+            # We take the corresonding choices and apply the `label` template
+            # Note: we do not allow `label` template to access other fields in multi-choice tasks
+            choices = [self._apply(c, templates.get("label", None), label=c) for c in data["choices"]]
+        # We first convert all fields except `input` to strings
+        transformed_data = copy.deepcopy(data)
+        for k, template in templates.items():
+            if k not in ["input", "verbalizer", "meta", "instruction", "label", "answer_context"]:
+                assert k in data, f"Template key `{k}` is not valid!"
+                transformed_data[k] = self._apply(data[k], template, **data)
+
+        # We then merge all other fields into the `input`
+        data["input"] = templates["input"].format(**transformed_data)
+        if "choices" in data.keys():
+            # Overwrite `choices` to the actual continuations
+            data["choices"] = choices
+        else:
+            # For generation tasks, we allow it to access to other stringified fields
+            kwargs = transformed_data
+            del kwargs["label"]
+            data["label"] = [self._apply(x, templates.get("label", None), **kwargs, label=x) for x in data["label"]]
+        return data
+
+    def _apply(self, data: Any, template: Optional[Any], **kwargs) -> str:
```

Comment:
You should do something like:

```python
# At the module level
TemplateData = Union[str, Dict[str, str], List[str], List[Dict[str, str]]]
Template = Union[str, Dict[str, str]]

# Here
def _apply(self, data: TemplateData, template: Optional[Template], **kwargs) -> str:
```

Also update the docstring to indicate that template can be `None`.

Reply:
Thanks for your suggestions! We have annotated types according to your advice, except that we use `RawData` instead of `TemplateData` as the data is expected not to be tied to a specific prompt template. We notice that the static analyzer cannot narrow `Union` down to a more specific type and raise errors. In this case, we use `# type: ignore` but add some assertions to guarantee the data type.